backends/graphics/atari/atari-graphics.cpp

```diff
@@ -0,0 +1,998 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "backends/graphics/atari/atari-graphics.h"
+
+#include <mint/cookie.h>
+#include <mint/falcon.h>
+#include <mint/osbind.h>
+
+#include "backends/graphics/atari/atari-graphics-asm.h"
+#include "backends/graphics/atari/videl-resolutions.h"
+#include "backends/keymapper/action.h"
+#include "backends/keymapper/keymap.h"
+
+#include "common/config-manager.h"
+#include "common/str.h"
+#include "common/textconsole.h"	// for warning() & error()
+#include "common/translation.h"
+#include "gui/ThemeEngine.h"
+
+#define SCREEN_ACTIVE
+
+AtariGraphicsManager::AtariGraphicsManager() {
+	_vgaMonitor = VgetMonitor() == MON_VGA;
+
+	// force standard GUI renderer (!DISABLE_FANCY_THEMES implies anti-aliased rendering in ThemeEngine.cpp)
+	ConfMan.set("gui_renderer", GUI::ThemeEngine::findModeConfigName(GUI::ThemeEngine::kGfxStandard));
+
+	g_system->getEventManager()->getEventDispatcher()->registerObserver(this, 10, false);
+}
+
+AtariGraphicsManager::~AtariGraphicsManager() {
+	g_system->getEventManager()->getEventDispatcher()->unregisterObserver(this);
+
+	for (int i = 0; i < SCREENS; ++i) {
+		Mfree(_screen[i]);
+		_screen[i] = _screenAligned[i] = nullptr;
+	}
+
+	Mfree(_overlayScreen);
+	_overlayScreen = nullptr;
+}
+
+bool AtariGraphicsManager::hasFeature(OSystem::Feature f) const {
+	switch (f) {
+	case OSystem::Feature::kFeatureAspectRatioCorrection:
+		debug("hasFeature(kFeatureAspectRatioCorrection): %d", !_vgaMonitor);
+		return !_vgaMonitor;
+	case OSystem::Feature::kFeatureCursorPalette:
+		debug("hasFeature(kFeatureCursorPalette): %d", isOverlayVisible());
+		return true;
+	case OSystem::Feature::kFeatureVSync:
+		debug("hasFeature(kFeatureVSync): %d", _vsync);
+		return true;
+	default:
+		return false;
+	}
+}
+
+void AtariGraphicsManager::setFeatureState(OSystem::Feature f, bool enable) {
+	switch (f) {
+	case OSystem::Feature::kFeatureAspectRatioCorrection:
+		debug("setFeatureState(kFeatureAspectRatioCorrection): %d", enable);
+		_oldAspectRatioCorrection = _aspectRatioCorrection;
+		_aspectRatioCorrection = enable;
+		break;
+	case OSystem::Feature::kFeatureVSync:
+		debug("setFeatureState(kFeatureVSync): %d", enable);
+		_vsync = enable;
+		break;
+	default:
+		[[fallthrough]];
+	}
+}
+
+bool AtariGraphicsManager::getFeatureState(OSystem::Feature f) const {
+	switch (f) {
+	case OSystem::Feature::kFeatureAspectRatioCorrection:
+		//debug("getFeatureState(kFeatureAspectRatioCorrection): %d", _aspectRatioCorrection);
+		return _aspectRatioCorrection;
+	case OSystem::Feature::kFeatureCursorPalette:
+		//debug("getFeatureState(kFeatureCursorPalette): %d", isOverlayVisible());
+		return isOverlayVisible();
```

Comment:
The main goal of this port is to support 8bpp games _only_. Therefore the cursor must be set by the engines, same as they did in the original code. The overlay can support cursor palettes is only thanks to the fact that Atari Falcon has RGB565 as one of its native resolutions so it would be shame not to use it (but it is also slower to render so that's why I support only 8bpp engines, so I can render them in 8bpp and not 16bpp).

Reply:
Understood, however, I am afraid, we were not quite consistent with checking this feature for the newer engines... Perhaps, more playtesting needs to be performed and those engines fixed. Maybe the Atari userbase could help us with determining that kind of problems.

Reply:
Yes, that's why I released 2.6.1 in the wild, to gather feedback on specialities like this. So far, nobody has complained about weird cursor colours.

Since there were times when ScummVM was 8bpp only, too I think that 8bpp engines should be fine. Maybe you have noticed, I'm logging usage of `hasFeature` for this very reason (btw the difference between `hasFeature` and `getFeatureState` would be another interesting discussion, usage of these two functions definitely calls for cleanup in engines).

So far all the games I have tested ask for the cursor palette for themselves, _while not being in the overlay_. So it's not like we show the overlay and suddenly some game gets an idea to check whether it shouldn't draw its game cursor.

Reply:
It's also worth pointing out that the Atari backend is currently the only one that doesn't support cursor palettes, so that's part of why so few engines check for it before using it. In particular, it'll be necessary for games that use standard Windows or Mac cursors, where the cursor palette is unlikely to resemble the screen palette.

Reply:
Actually I have taken your recommendation to take a look into that RGB332 mode into account. It's a neat and elegant solution, just tested it. So this will all go away in my next PR, the backend will become 100% 8bpp only.

Reply:
@ccawley2011 hmm, as KolibriOS uses the SDL backend, I can't easily see how it is solved in there but to me it looks like it also can't support custom cursor palettes as it uses 8bpp rendering? (not to speak about the overlay in the neat RGB332 mode)

Reply:
Converting between palette formats is handled by `SDL_BlitSurface`. It shouldn't be too difficult to add a simple conversion function to `graphics/blit.h` and call it whenever the palette changes, though.

Reply:
I see. In that case I am indeed going to wait for feedback from Atari users to get an idea how big issue this is during real usage.

I'd like to avoid as many on the fly conversions as possible; currently many games call `setMouseCursor` way too often. What makes me sad is that it is often because of cursor fading in/out, i.e. something which could be solved via usage of `setCursorPalette` but nope, they just supply new pixels values (even if the cursor feature is enabled) :(

So my motivation to support cursor palette in 8bpp is pretty low currently.